Framework/Core/src/WorkflowHelpers.cxx

```diff
@@ -1233,4 +1233,30 @@ std::vector<InputSpec> WorkflowHelpers::computeDanglingOutputs(WorkflowSpec cons
 }
 
 #pragma diagnostic pop
+void WorkflowHelpers::validateEdges(WorkflowSpec const& workflow,
+                                    std::vector<DeviceConnectionEdge> const& edges,
+                                    std::vector<OutputSpec> const& outputs)
+{
+  // Iterate over all the edges.
+  // Get the input lifetime and the output lifetime.
+  // Output lifetime must be Timeframe if the input lifetime is Timeframe.
+  std::string errors;
+  for (auto& edge : edges) {
+    auto& producer = workflow[edge.producer];
+    auto& consumer = workflow[edge.consumer];
+    auto& output = outputs[edge.outputGlobalIndex];
+    auto& input = consumer.inputs[edge.consumerInputIndex];
+    // We cannot simply say "!Timeframe" because we want to allow
+    // e.g. Enumeration => Timeframe
+    if (input.lifetime == Lifetime::Timeframe && output.lifetime == Lifetime::Sporadic) {
```

Comment:
shouldn't this be input.lifetime == Lifetime::Timeframe && output.lifetime != Lifetime::Timeframe?

Reply:
See the comment above. Lifetime::Enumeration outputs are compatible with Lifetime::Timeframe (and so is Lifetime::Timer). I would start with this.

Reply:
Hm, that makes it more complicated of course. Though I don't understand why in that case the input is not also enumaration or timer?

In any case, what about lifetime::optional? I assume this we should then also check?

Reply:
The input is not an enumeration / timer, because by design the destinations do not know how their inputs are created.

For the second point, if something is optional it will always be there, deadbeef or not.

Reply:
Don't understand, what does this have to do with 0xDEADBEEF?
Isn't optional the same as sporadic?
If not, and if optional is there only for the DEADBEEF mechanism and optional is supposed to be always there, then I think we got rid of the DEADBEEF mechanism now, and we should remove optional entirely?

Reply:
No, Optional is something which results in "DEADBEEF" if not there by the time the last timeframe is there. Sporadic is simply some data which might or might not be there at each iteration, but it does not result in DEADBEEF. If you prefer, Optional is only for inputs, sporadic is for both inputs and outputs. The real confusion I think comes from the fact that the same lifetime terms are used by inputs and outputs, while in reality they might indicate two different things.